Task object

norkunas opened the pull request:
# Pull Request

## Related issue
Fixes #703

## What does this PR do?
- Bumps PHP requirement to 8.1;
- Adds missing types where there were none at all, or was impossible to use because of unions;
- Adds Task object for type safety to be returned instead of undocumented array shapes;
- Fully enables `trailing_comma_in_multiline` cs fixer rule;
- Disables `php_unit_strict` cs fixer rule, because it breaks tests; 

**This introduces a BC break for someone who is extending meilisearch classes, but not for those who consumes the data, as Task object was added implementing `\ArrayAccess`, so all raw data can still be accessed.**

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

tacman commented:
YAY!

If these are to be deprecated, are they going to be replaced with direct access to the object?

```php
    public function getData(): array
    {
        return $this->data;
    }

    // @todo: deprecate
    public function offsetExists(mixed $offset): bool
    {
        return \array_key_exists($offset, $this->data);
    }

    // @todo: deprecate
    public function offsetGet(mixed $offset): mixed
    {
        return $this->data[$offset] ?? null;
    }

    // @todo: deprecate
    public function offsetSet(mixed $offset, mixed $value): void
    {
        throw new \LogicException(\sprintf('Setting data on "%s::%s" is not supported.', get_debug_type($this), $offset));
    }

    // @todo: deprecate
    public function offsetUnset(mixed $offset): void
    {
        throw new \LogicException(\sprintf('Unsetting data on "%s::%s" is not supported.', get_debug_type($this), $offset));
    }
```

norkunas commented:
@tacman if this PR will get merged, then you'll get access to both ways: `$task->getError()` or `$task['error']`
but imho previous way should be deprecated, because it does not provide types

Strift commented:
Hi, thanks for your PR. 

Do we want to release this in a v2.x? 

Can you add a "migration guide" to your first comment to highlight the required implementation changes?

norkunas commented:
@Strift

> Do we want to release this in a v2.x?

Based on semver, it should be in v2.x, but if you want to be in line with [meilisearch/meilisearch](https://github.com/meilisearch/meilisearch) repository, I think it's reasonable to release this in v1.

> Can you add a "migration guide" to your first comment to highlight the required implementation changes?

Can you link to some previous comment where I could see how this looks like?

Strift commented:
Hi @norkunas, no need to follow a specific format. A simple before/after of the relevant parts of the code is enough. Here's an example for meilisearch JS: https://github.com/meilisearch/meilisearch-js/releases/tag/v0.48.0 

It's mainly so I have examples to refer to when writing the release changelog. Double-checking with the actual code changes is better, but the code samples are a good resource too for me, especially if I write the changelog a long time after reviewing the PR.

norkunas commented:
@Strift I've updated description, is it ok for you?
Also it should be decided if we deprecate the array access or not.

norkunas commented:
I also have a side idea for another improvement:
Add `wait(int $timeoutInMs, int $intervalInMs)` method to task.
So instead of:
```php
$task = $this->client->createIndex('new-index', ['primaryKey' => 'objectID']);
$task = $this->index->waitForTask($task->getTaskUid());
```

we could write:
```php
$task = $this->client->createIndex('new-index', ['primaryKey' => 'objectID'])->wait();
```

But not sure, as then we'd need for this method to work at least inject in in the Task object the `Meilisearch\Endpoints\Tasks` endpoint, or the http client, or the meilisearch client itself..

norkunas commented:
Hi, I'm for merging to v1 :) I think this is an acceptable BC break to provide type safety

norkunas commented:
I also think that it would be good to have typed task `details` based on https://www.meilisearch.com/docs/reference/api/tasks#details, because now it's always `array<mixed>`, so static analyzers will always complain.

norkunas commented on src/Endpoints/Delegates/HandlesSystem.php:
```diff
@@ -47,7 +48,7 @@ public function generateTenantToken(string $apiKeyUid, $searchRules, array $opti
     }
 
-    public function swapIndexes(array $indexes): array
+    public function swapIndexes(array $indexes): Task
```
@Copilot but it shouldn't, because `Indexes::swapIndexes` wraps it

Strift commented on tests/Settings/PaginationTest.php:
```diff
@@ -0,0 +1,51 @@
+        self::assertSame(['maxTotalHits' => 100], $this->index->getPagination());
+    }
+
+    public function testResetRankingRules(): void
```
```suggestion
    public function testResetPagination(): void
```

Strift reviewed (changes requested):
This mostly looks good to me. I left a few comments.

I would love to release this with the changes related to Meilisearch v1.15.

Also, Codecov says the following lines are not covered by tests:
- src/Contracts/Task.php#L114
- src/Contracts/Task.php#L116

Strift commented:
> I also have a side idea for another improvement:
> Add wait(int $timeoutInMs, int $intervalInMs) method to task.

This could be a welcome addition, but I prefer having it in a separate PR to keep this one's scope minimal.

> I also think that it would be good to have typed task details based on https://www.meilisearch.com/docs/reference/api/tasks#details, because now it's always array<mixed>, so static analyzers will always complain.

Can we tackle this in another PR? We can create an issue to track it in the meantime.

norkunas commented:
Thanks @Strift for review, I'm currently preparing for `details` data to be typed.

norkunas commented on src/Contracts/TaskType.php:
```diff
@@ -0,0 +1,21 @@
+    case IndexSwap = 'indexSwap';
+    case DocumentAdditionOrUpdate = 'documentAdditionOrUpdate';
+    case DocumentDeletion = 'documentDeletion';
+    case DocumentEdition = 'documentEdition';
```
Will remove

norkunas replied:
Oh now I see that it was used in tests which broke after I removed it.

norkunas replied:
@Strift as you can see meilisearch docs are outdated: https://github.com/meilisearch/meilisearch-php/actions/runs/15012425388/job/42183478341?pr=735

Strift replied:
I will report it, thanks!

norkunas commented on src/Endpoints/Delegates/HandlesDocuments.php:
```diff
@@ -6,15 +6,14 @@
-    public function getDocument($documentId, ?array $fields = null)
+    public function getDocument(string|int $documentId, ?array $fields = null): array
     {
-        $this->assertValidDocumentId($documentId);
```
Forgot to add `@param non-empty-string` :)

norkunas commented on tests/Contracts/TaskTest.php:
```diff
@@ -0,0 +1,112 @@
+            'type' => 'invalid_request',
+            'link' => 'https://docs.meilisearch.com/errors#index_not_found',
+        ], $task->getError());
+
```
I don't think it's worth to move. Either we'll remove the test later or just this part from current test, so it's the same

norkunas commented on src/Contracts/TaskDetails/IndexUpdateDetails.php:
```diff
@@ -0,0 +1,30 @@
+    ) {
+    }
+
+    public static function fromArray(array $data): self
+    {
+        return new self(
+            $data['primaryKey'],
+        );
+    }
```
Why? it's not optional field

norkunas commented on src/Contracts/TaskDetails/IndexCreationDetails.php:
```diff
@@ -0,0 +1,30 @@
+    ) {
+    }
+
+    public static function fromArray(array $data): self
+    {
+        return new self(
+            $data['primaryKey'],
+        );
+    }
```
it's always available

brunoocasali commented on .github/workflows/pre-release-tests.yml:
```diff
@@ -88,39 +84,3 @@ jobs:
       - name: Run test suite
         run: sh scripts/tests.sh
-
-  test_php_7_guzzle_6:
-    runs-on: ubuntu-latest
-    needs: ['meilisearch-version']
-    name: integration-tests-against-rc (PHP 7.4 & Guzzle 6)
-    services:
-      meilisearch:
-        image: getmeili/meilisearch:${{ needs.meilisearch-version.outputs.version }}
-        env:
-          MEILI_MASTER_KEY: 'masterKey'
-          MEILI_NO_ANALYTICS: 'true'
-        ports:
-          - '7700:7700'
-
-    steps:
-      - name: Checkout code
-        uses: actions/checkout@v4
-
-      - name: Install PHP
-        uses: shivammathur/setup-php@v2
-        with:
-          php-version: '7.4'
-          coverage: none
-
-      - name: Validate composer.json and composer.lock
-        run: composer validate
-
-      - name: Switch to Guzzle 6
-        run: |
-          sed -i 's/"guzzlehttp\/guzzle": "^[0-9]\+\.[0-9]\+\.[0-9]\+\",/"php-http\/guzzle6-adapter": "^2.0.2",/' composer.json
-
-      - name: Install dependencies
-        uses: ramsey/composer-install@v3
-
-      - name: Run test suite - php-http/guzzle6-adapter
-        run: sh scripts/tests.sh
```
OMG thats beautiful 👏👏

brunoocasali commented on .php-cs-fixer.dist.php:
```diff
@@ -15,9 +15,8 @@
-        'php_unit_strict' => true,
-        // @todo: when we'll support only PHP 8.0 and upper, we can enable `parameters` for `trailing_comma_in_multiline` rule
-        'trailing_comma_in_multiline' => ['after_heredoc' => true, 'elements' => ['array_destructuring', 'arrays', 'match'/* , 'parameters' */]],
+        'php_unit_strict' => false,
```
Why changing this?

norkunas replied:
Because it converts some call from `self::assertEquals` to `self::assertSame` and then the tests break

brunoocasali commented on src/Contracts/Task.php:
```diff
@@ -0,0 +1,197 @@
+        private readonly ?string $duration = null,
+        private readonly ?int $canceledBy = null,
+        private readonly ?int $batchUid = null,
+        private readonly ?TaskDetails $details = null,
```
When we introduce a new task type in the engine does will this code break?

norkunas replied:
Yes, this lib must be prepared to accept new task types/statuses/task details on new meilisearch releases, same as with other features, that @Strift or others adds when meilisearch gets them :)

brunoocasali replied:
But when I mean "break," will the user really see a parsing error or something like this if the type is not supported?

brunoocasali replied:
I'm unsure if this is enough when we decided internally to move Meilisearch (the engine) to v1. One of the requirements for the integrations would be to be able to keep updating the engine no matter which version you have in the SDK.

So far, we've been successful with that promise if I'm not mistaken 😅. Would it be possible to handle it in a way we can redirect the code to a non-typed task type? So if a user updates Meilisearch, it won't break for them, and to get the full capabilities from the SDK, it will just require them to later upgrade the SDK, WDYT?

norkunas replied:
But any change in api could break this sdk, so..

norkunas replied:
> One of the requirements for the integrations would be to be able to keep updating the engine no matter which version you have in the SDK.

if that's a requirement, then we can resort back to using strings, which is sad, but that's your decision

brunoocasali commented on tests/Endpoints/BatchesTest.php:
```diff
@@ -36,7 +36,7 @@ public function testGetAllBatchesWithIndexUidFilters(): void
     public function testGetAllBatchesWithTasksFilters(): void
     {
         $tasks = $this->client->getTasks(new TasksQuery())->getResults();
-        $response = $this->client->getBatches((new BatchesQuery())->setUids([$tasks[0]['uid']]));
```
Although now this is an object since the task implements `ArrayAccess`, I can still do the `['uid']` right?

norkunas replied:
Yup, but now totally removed

brunoocasali commented on tests/Settings/SearchableAttributesTest.php:
```diff
@@ -28,26 +28,19 @@ public function testUpdateSearchableAttributes(): void
         $index = $this->createEmptyIndex($this->safeIndexName('books-1'));
-        $promise = $index->resetSearchableAttributes();
-
-        $this->assertIsValidPromise($promise);
```
Thanks for removing this helper method 🙏

brunoocasali commented:
@norkunas After further discussing the subject with @curquiza, we decided it would be more prudent to release it as v2 since it introduces a few breaking changes, especially the deprecation of PHP 7.

If that becomes a burden for our users later, we can introduce a table to showcase which SDK version supports which Meilisearch version. In any case, I think that's better, the engine and their SDKs are independent pieces of software that should be able to evolve independently and should remain adherent to semver rules :) 

So, since we are doing a v2, feel free to suggest other potential breaking changes so we can release it in the same package!

Thanks a lot for the PR by the way, loved it! 

CC: @Strift

norkunas commented:
> @norkunas After further discussing the subject with @curquiza, we decided it would be more prudent to release it as v2 since it introduces a few breaking changes, especially the deprecation of PHP 7.
> 
> If that becomes a burden for our users later, we can introduce a table to showcase which SDK version supports which Meilisearch version. In any case, I think that's better, the engine and their SDKs are independent pieces of software that should be able to evolve independently and should remain adherent to semver rules :) 
> 
> So, since we are doing a v2, feel free to suggest other potential breaking changes so we can release it in the same package!
> 
> Thanks a lot for the PR by the way, loved it! 
> 
> CC: @Strift 

Sounds good 😉

norkunas commented:
Then I think we can remove array access of task data here?

norkunas commented on src/Contracts/TaskDetails/TaskDeletionDetails.php:
```diff
@@ -0,0 +1,38 @@
+    ) {
+    }
+
+    public static function fromArray(array $data): self
+    {
+        return new self(
+            $data['matchedTasks'],
+            $data['deletedTasks'],
+            $data['originalFilter'],
+        );
+    }
```
These keys are always available

brunoocasali commented:
> Then I think we can remove array access of task data here?

If you think we have more benefits by removing it than by keeping, I agree in removing it.

norkunas commented:
@Strift i should switch to v2 branch target here?

Strift commented:
Hi @norkunas, yes, please. Let's target the `v2` branch here.

norkunas commented:
@brunoocasali that's not up to me. v2 needs to be upmerged, because `main` has some pushed changes

Strift commented:
Apologies, I may have made a mistake.

I merged the changes from v2 into this branch, but maybe I should have rebased the PR instead.

norkunas commented:
I never do merges

norkunas commented:
@Strift i've removed those commits. i can now rebase, but the PR target is still v2, i should change it to main, which is now v2, right?

Strift commented on src/functions.php:
```diff
@@ -0,0 +1,15 @@
+
+/**
+ * @internal
+ */
```
```suggestion
/**
 * @internal
 * 
 * Creates a partially applied function by binding initial arguments to the given callable. 
 * 
 * Returns a Closure that, when invoked, calls the original callable with the bound arguments prepended
 * to any new ones. 
 * 
 * Used internally to build reusable “waiter” functions (e.g., binding the HTTP client to 
 * task-waiting logic) and reduce repetitive argument passing.
 */
```

Strift commented on tests/Endpoints/IndexTest.php:
```diff
@@ -185,56 +186,47 @@ public function testGetTasks(): void
+        $task = $this->index->addDocuments([['id' => 1, 'title' => 'Pride and Prejudice']])->wait(750, 20);
+
+        self::assertSame(TaskStatus::Succeeded, $task->getStatus());
+        self::assertSame($task->getTaskUid(), $task->getTaskUid());
```
```suggestion
```

Strift commented on tests/Endpoints/IndexTest.php:
```diff
@@ -185,56 +186,47 @@ public function testGetTasks(): void
+        $task = $this->index->addDocuments([['id' => 1, 'title' => 'Pride and Prejudice']])->wait(1000);
+
+        self::assertSame(TaskStatus::Succeeded, $task->getStatus());
+        self::assertSame($task->getTaskUid(), $task->getTaskUid());
```
```suggestion
```

Strift commented on tests/Endpoints/IndexTest.php:
```diff
@@ -184,56 +186,47 @@ public function testGetTasks(): void
+        $task = $this->index->addDocuments([['id' => 1, 'title' => 'Pride and Prejudice']])->wait();
+
+        self::assertSame(TaskStatus::Succeeded, $task->getStatus());
+        self::assertSame($task->getTaskUid(), $task->getTaskUid());
```
@norkunas I agree and would remove this assertion

```suggestion
```

Strift commented on tests/Endpoints/DocumentsTest.php:
```diff
@@ -53,10 +53,7 @@ public function testAddDocumentWithSpecialChars(): void
-
-        $this->assertIsValidPromise($promise);
-        $index->waitForTask($promise['taskUid']);
+        $task = $index->addDocuments($documents)->wait();
```
```suggestion
        $index->addDocuments($documents)->wait();
```

Strift commented on tests/Endpoints/DocumentsTest.php:
```diff
@@ -298,9 +280,10 @@ public function testUpdateDocumentsByFunction(): void
-        $documentPromise = $index->addDocuments(self::DOCUMENTS);
-        $index->waitForTask($documentPromise['taskUid']);
+
+        $task = $index->addDocuments(self::DOCUMENTS)->wait();
```
```suggestion
        $index->addDocuments(self::DOCUMENTS)->wait();
```

norkunas commented:
I'll try to find some time to continue working on this, but currently my primary work consumes all my time

norkunas commented on src/Contracts/SimilarDocumentsQuery.php:
```diff
@@ -7,9 +7,9 @@
+     * @var int|non-empty-string
      */
-    private $id;
+    private int|string $id;
```
This is a contract, if someone does not follow, then it's on them. Static analysis should warn them

norkunas commented:
@Strift enums basically are already value objects, so I don't see a point in wrapping them.
I've added `Unknown` values to enums itself. Can you please recheck the PR? I've rebased to fix conflicts (oh boy, it took some time.. :D), i hope this will be okay for you

norkunas commented:
Btw `integration-tests (PHP 8.4)` and ``integration-tests (PHP 8.0)` should be removed from repo settings as required jobs.

Strift reviewed (approved):
Great work! All things LGTM now.

I have a question regarding the parsing of task types/statuses. Should we add a PSR-3 warning? 

e.g. `"Unknown Meilisearch task status: <status>; mapped to TaskStatus::Unknown. Consider upgrading the SDK for typed support."`
